File: python/easy/1512_Number_of_Good_Pairs.py
from typing import *
class Solution:
    # 36 ms, faster than 38.77% of Python3 online submissions for Number of Good Pairs
    # 14.3 MB, less than 41.67% of Python3 online submissions for Number of Good Pairs
    def numIdenticalPairs(self, nums: List[int]) -> int:
        ans = 0
        for idx1, num1 in enumerate(nums[:-1]):
            for idx2,num2 in enumerate(nums[idx1+1:]):
                if num1 == num2:
                    ans+=1
        return ans

    #  28 ms, faster than 90.91% of Python3 online submissions for Number of Good Pairs
    # 14.2 MB, less than 41.67% of Python3 online submissions for Number of Good Pairs.
    def numIdenticalPairs(self, nums: List[int]) -> int:
        from collections import defaultdict
        numCount = defaultdict(int)
        for x in nums:
            numCount[x] += 1
        ans = 0
        for count in numCount.values():
            if count >= 2:
                ans += (count * (count-1)) // 2
        return ans

if __name__ == "__main__":
    so = Solution()
    print(so.numIdenticalPairs([1,1,1,1]))

File: c/easy/1512_Number_of_Good_Pairs.c
#include <stdio.h>
#include <stdlib.h>

int numIdenticalPairs(int* nums, int numsSize);

int main(void){
    int nums[] = {4,4,2,2};
    printf("%d\n", numIdenticalPairs(nums, sizeof(nums)/sizeof(int)));
}


int numIdenticalPairs(int* nums, int numsSize){
    int * dict = (int *) malloc(sizeof(int) * 100);
    int i;
    int ans = 0;

    for (i = 0; i < numsSize; i++){
        dict[i] = 0;
    }
    for (i = 0; i < numsSize; i++){
        dict[nums[i]] += 1;
        printf("nums[%d] = %d, dict[%d] = %d\n", i, nums[i], nums[i], dict[i]);
    }
    for (i = 0; i < 100; i++){
        printf("dict[%d] = %d\n", i, dict[i]);
        if (dict[i] >=2 ){
            ans += dict[i] * (dict[i] -1 ) / 2;
        }
    }
    return ans;
}